File: src/ts/src/modes/apply.ts
/** 공통 치환 유틸리티 — DetectionResult 리스트로 문자열을 안전하게 치환. */
import { resolveOverlaps } from "../core/overlap.js";
import type { DetectionResult } from "../core/types.js";

/** span 을 대체 문자열로 바꾸는 콜백 — Python 의 ``replacer`` 대응. */
export type Replacer = (d: DetectionResult) => string;

/**
 * Apply ``replacer`` to each (deduped, sorted) detection span in *text*.
 *
 * 겹침 해소는 core/overlap 단일 구현을 dedup 으로 재사용한다 (detect_all 과
 * 동일 우선순위 — Python 이 resolve_overlaps 를 재사용하는 구조 그대로).
 */
export function applySubstitutions(
  text: string,
  detections: Iterable<DetectionResult>,
  replacer: Replacer,
): string {
  const ordered = resolveOverlaps([...detections]);
  if (ordered.length === 0) return text;
  const pieces: string[] = [];
  let cursor = 0;
  for (const d of ordered) {
    if (d.start > cursor) pieces.push(text.slice(cursor, d.start));
    pieces.push(replacer(d));
    cursor = d.end;
  }
  pieces.push(text.slice(cursor));
  return pieces.join("");
}

File: src/python/ko_pii/modes/_apply.py
"""공통 치환 유틸리티 — DetectionResult 리스트로 문자열을 안전하게 치환."""
from __future__ import annotations

from typing import Callable, Iterable

from ko_pii.core.types import DetectionResult
# 겹침 해소는 core.overlap 단일 구현 사용 (detect_all 과 동일 우선순위).
from ko_pii.core.overlap import resolve_overlaps as _dedup_and_sort


def apply_substitutions(
    text: str,
    detections: Iterable[DetectionResult],
    replacer: Callable[[DetectionResult], str],
) -> str:
    """Apply ``replacer`` to each (deduped, sorted) detection span in *text*."""
    ordered = _dedup_and_sort(detections)
    if not ordered:
        return text
    pieces: list[str] = []
    cursor = 0
    for d in ordered:
        if d.start > cursor:
            pieces.append(text[cursor:d.start])
        pieces.append(replacer(d))
        cursor = d.end
    pieces.append(text[cursor:])
    return "".join(pieces)
